### PR TITLE
Extract functions for service worker usage, and add initial MSC3916 playwright test (when supported)

### PR DESCRIPTION
**Pull request by turt2live**
For https://github.com/element-hq/element-web/pull/27326

See https://github.com/matrix-org/matrix-react-sdk/pull/12414#issuecomment-2072424525 for context

## Checklist

-   [x] Tests written for new code (and old code if feasible).
-   [x] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [x] Linter and other CI checks pass.
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md)).

### CODE REVIEW

**Comment by turt2live** [comment]
As noted in the diff, tests for this are effectively impossible. See https://github.com/microsoft/playwright/issues/15684#issuecomment-2070862042 for a bit more detail.

**Comment by turt2live** [comment]
SonarCloud appears to be complaining about code which is tested implicitly by Playwright tests. The functions have been refactored here rather than being newly introduced.

**Comment by turt2live** [comment]
End to end test failure looks like a flake of some kind (cause appears unrelated to this PR's changes).

File: src/utils/tokens/pickling.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        additionalData[userId.length + 1 + i] = deviceId.charCodeAt(i);
+    }
+    return additionalData;
+}
+
+/**
+ * Assuming `data` is fully provided, this converts the data into a usable key then base64 encodes it for use in
+ * other functions. `data` may be partially or completely undefined to deduplicate some validation logic shared
+ * across callers - when undefined in part or in full, this function returns undefined too. If crypto functions
+ * are not available, this function returns undefined regardless of input.
```
Comment (richvdh) [inline src/utils/tokens/pickling.ts:47]: I think it's important to start documentation with a single-sentence summary of what the method does. Which this kinda does, but if I only read the first sentence, I'm kinda left expecting that the function will do something :sparkles:magical:sparkles: to populate the data. (Also, that first sentence should be a paragraph on its own).

Also also: surprised not to see the word "decrypt" in this description.

```suggestion
 * Decrypts the provided data into a pickle key and base64-encodes it ready for use elsewhere.
 *
 * If `data` is undefined in part or in full, returns undefined.
 *
 *  If crypto functions are not available, returns undefined regardless of input.
```

File: src/utils/tokens/pickling.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * @param data An object containing the pickle key data: encrypted payload, initialization vector (IV), and crypto key.
```
Comment (richvdh) [inline src/utils/tokens/pickling.ts:49]:
```suggestion
 * @param data An object containing the encrypted pickle key data: encrypted payload, initialization vector (IV), and crypto key. Typically loaded from indexedDB.
```

**Review by richvdh — CHANGES_REQUESTED** [review]
I'm slightly confused what's going on here. I see lots of factoring out to new files, which is great, but I don't think the methods are being used anywhere new? So is this just groundwork for future changes? (In other words: how does this actually help with service workers?)

File: src/utils/StorageAccess.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+ */
+export function getIndexedDb(): IDBFactory | undefined {
+    // IndexedDB loading is lazy for easier testing.
+
+    // just *accessing* _indexedDB throws an exception in firefox with
+    // indexeddb disabled.
+    try {
+        // `self` is preferred for service workers, which access this file's functions.
+        // We check `self` first because `window` returns something which doesn't work for service workers.
+        return self?.indexedDB ? self.indexedDB : window.indexedDB;
```
[the remark this answers is not in the document]
Comment (turt2live) [reply src/utils/StorageAccess.ts:30]: For reasons I have not yet been able to comprehend, this does not work. It needs to be a ternary at best. 

I'm not sure why. I agree it shouldn't matter.

**Comment by turt2live** [comment]
> I'm slightly confused what's going on here. I see lots of factoring out to new files, which is great, but I don't think the methods are being used anywhere new? So is this just groundwork for future changes? (In other words: how does this actually help with service workers?)

This refactoring is for the service worker at the element-web layer. The functions are used there, and not splitting them out would mean imports which bring in half the react-sdk. Those imports include things which cannot run in a service worker, namely localstorage access. 

This PR splits out the minimum required functions to avoid bringing in stacks upon stacks of error-inducing code.

**Comment by richvdh** [comment]
> This refactoring is for the service worker at the element-web layer. The functions are used there, and not splitting them out would mean imports which bring in half the react-sdk. Those imports include things which cannot run in a service worker, namely localstorage access.

Yes, I've now read https://github.com/element-hq/element-web/pull/27326. Sorry, should have checked that sooner!

File: src/utils/StorageAccess.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply src/utils/StorageAccess.ts:30]: [is self.indexedDB falsy but not-nullish, maybe? In which case maybe `self?.indexedDB || window.indexedDB` would do the right thing? Not that that is necessarily better than the ternary.]

[the remark this answers is not in the document]
Comment (richvdh) [reply src/utils/StorageAccess.ts:30]: `||` does not convert anything into a boolean in JS. But I agree that's not completely obvious to the casual reader, hence "Not that that is necessarily better than the ternary." I don't have a strong preference for `||` vs `?:`; I'm just musing about the problem.

File: src/utils/StorageAccess.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    } catch (e) {}
+}
+
+/* Simple wrapper functions around IndexedDB.
+ */
```
[the remark this answers is not in the document]
Comment (turt2live) [reply src/utils/StorageAccess.ts:35]: This is a copy/paste artifact. I don't think it provides value anymore - will remove.

File: src/utils/tokens/pickling.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { encodeUnpaddedBase64 } from "matrix-js-sdk/src/matrix";
+import { logger } from "matrix-js-sdk/src/logger";
+
+/**
+ * Retrieves additional data for a pickle key (used to decrypt access tokens and similar
+ * stored in IndexedDB).
```
[the remark this answers is not in the document]
Comment (turt2live) [reply src/utils/tokens/pickling.ts:25]: It is the cryptographic sense: authenticated, but not encrypted. I can't really speak to *why* we do this, or if it's a good idea, but I have added a bunch of documentation around this particular area in hopes of someone else improving it.

**Review by richvdh — APPROVED** [review]
lgtm

**Pull request merged**